### DISCUSSION TITLE
How to perform identical transform on two different Images?

### DISCUSSION THREAD

**Discussion by AhmedHussKhalifa** [yoshitomo-matsubara/torchdistill#171] [Q&A]
Hey,
I am trying to apply the same transformation on two different images using BaseDatasetWrappe. I hard-coded the default transformation using the down-mentioned function (transform_new) in  [BaseDatasetWrapper](https://github.com/yoshitomo-matsubara/torchdistill/blob/de9cd86d392d28c12ce654d0efd7234a2ab15435/torchdistill/datasets/wrapper.py#L24) using the same parameter you have provided in yaml file.  I can not reach the same performance as the default yaml file. I am using the yaml file transformation as the default check to observe if I am implementing things in the right way or not. I do not know what is wrong with my implementation. Could you let me know what is wrong with my implementation?

I am using this function as a start point to scale things up to apply the same transformation on two different images.

    'def transform_new(self, image):
        # Resize
        resize = transforms.Resize(size=(224, 224))
        image = resize(image)

        # Random crop
        i, j, h, w = transforms.RandomCrop.get_params(
            image, output_size=(224, 224))
        image = TF.crop(image, i, j, h, w)

        # Random horizontal flipping
        if random.random() > 0.5:
            image = TF.hflip(image)

        # Random vertical flipping
        # if random.random() > 0.5:
        #     image = TF.vflip(image)

        # Transform to tensor
        image = TF.to_tensor(image)

        norm = transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225))
        image = norm(image)

        return image`

**Comment by yoshitomo-matsubara** [comment]
Hi @AhmedHussKhalifa ,

A few questions about your implementation:
1. What do you mean by "two different images" ? I don't understand what you want to do based on this code
2. What do you import as `TF`?
3. Did you confirm that your hard-coded `transform_new` function is called when executing the code?
4. Most importantly, if you hard-code and call the transform function in `BaseDatasetWrapper`, you must have use two transforms functions (1st time: [`self.org_dataset.__get_item__(index)`](https://github.com/yoshitomo-matsubara/torchdistill/blob/de9cd86d392d28c12ce654d0efd7234a2ab15435/torchdistill/datasets/wrapper.py#L29), 2nd time: your hard-coded transform) as `self.org_dataset` does transform the image. e.g., Check [CIFAR-10's `__getitem__` function](https://github.com/pytorch/vision/blob/master/torchvision/datasets/cifar.py#L106-L126)

**Reply by AhmedHussKhalifa** [reply]
Hey @yoshitomo-matsubara,

Thank you for your quick reply.

> 1. What do you mean by "two different images" ? I don't understand what you want to do based on this code

I meant if I want to make the same data augmentation to 2 different loaded images at  `BaseDatasetWrapper`.

> 2. What do you import as TF?
Sorry, I forgot to mention it.

`import.transforms.functional as TF` 

> 3. I think that resizing and then random cropping is not the right way of implementing the default way as shown here.
I think I have to remove the resize part of the `transform_new` function, **Right?**

    'def transform_new(self, image):
        # Resize
        resize = transforms.Resize(size=(224, 224))
        image = resize(image)

        # Random crop
        i, j, h, w = transforms.RandomCrop.get_params(
            image, output_size=(224, 224))
        image = TF.crop(image, i, j, h, w)'

> 4. Most importantly, if you hard-code and call the transform function in BaseDatasetWrapper, you must have use two transforms functions (1st time: self.org_dataset.__get_item__(index), 2nd time: your hard-coded transform) as self.org_dataset does transform the image. e.g., Check CIFAR-10's __getitem__ function.


I have actually use the same way of implementing  `__getitem__ ` here[CIFAR-10's __getitem__ function](https://github.com/pytorch/vision/blob/972ca657416c5896ba6e0f5fe7c0d0f3e99e1087/torchvision/datasets/cifar.py#L106-L126) where I created my own `BaseDatasetWrapper`.

I have another question, how do save the logits of the augmented images from the teacher in [ CacheableDataset](https://github.com/yoshitomo-matsubara/torchdistill/blob/de9cd86d392d28c12ce654d0efd7234a2ab15435/torchdistill/datasets/wrapper.py#L36) while it should be randomly augmented?

**Reply by yoshitomo-matsubara** [reply]
Hi @AhmedHussKhalifa 

> I meant if I want to make the same data augmentation to 2 different loaded images at BaseDatasetWrapper.

I still don't understand, specifically the difference from the data augmentation defined in my yaml files. Isn't it applying different transforms to some portion in batch (say 16 of 32 samples in a batch) randomly like with 50% chance? 

> I think that resizing and then random cropping is not the right way of implementing the default way as shown here.
I think I have to remove the resize part of the transform_new function, Right?

> I have actually use the same way of implementing __getitem__ hereCIFAR-10's getitem function where I created my own BaseDatasetWrapper.

Can you share the code of your hard-coded BaseDatasetWrapper? Though I don't recommend hard-coding them in the existing modules in torchdistill, I cannot understand the above two points without seeing your code.

> I have another question, how do save the logits of the augmented images from the teacher in CacheableDataset while it should be randomly augmented?

If you use data augmentation, `CacheableDataset` won't help as it saves the (intermediate) outputs from teachers in the 1st epoch using sample indices as keys.
Since every epoch you will have different inputs even for the same sample indices, it doesn't make sense to cache the outputs from teacher.

**Reply by AhmedHussKhalifa** [reply]
Hey,
Sorry for not clearing things right away. I have implemented it in this way just to fix a certain data augmentation effects and oberserve the varition of some adversarial attacks on the input image.

    ` def apply_transform( image, aug_parameters):

        i, j, h, w, rand_HorFlip = aug_parameters
        image  = TF.resized_crop(image, i, j, h, w, size=(224,224), interpolation=TF.InterpolationMode.BILINEAR)

        if rand_HorFlip > 0.5:
            image = TF.hflip(image)
    
        # Transform to tensor
        image = TF.to_tensor(image)

        norm = transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225))
        image = norm(image)
        return image `

I have another question regarding the KD framework. If we have an overfit techer model, would we expect that it would not help in training the student as it provid roughly the hot encoding. Therfore, the propagated loss will result as an average of 2 hot encoders.

**Reply by yoshitomo-matsubara** [reply]
Hi @AhmedHussKhalifa 

The code still looks almost the same with your first code. As said in "Can you share the code of your hard-coded BaseDatasetWrapper?", I requested **your hard-coded BaseDatasetWrapper** (not your transform function). If you still see some performance degradation by your hard-coded transform, my doubt is on **your hard-coded BaseDatasetWrapper** rather than the transform function you shared.

> I have another question regarding the KD framework. If we have an overfit techer model, would we expect that it would not help in training the student as it provid roughly the hot encoding. Therfore, the propagated loss will result as an average of 2 hot encoders.

If by "a overfit teacher model" you mean the teacher model is too confident in its prediction (e.g., 0.999 for a category, and almost zero for all the other categories given an input), it won't help student model train well since it's almost the same as the true label (one-hot vector). There are some papers discussing such effects of over-confident teacher models in knowledge distillation.

**Comment by AhmedHussKhalifa** [comment]
Hey @yoshitomo-matsubara 

I have a following up question.

Let's assume that I have an initialized object for an attack method in ./example/imageclassification.py, and I want to utilize it in BaseDatasetWrapper(). what changes would need to be made to BaseDatasetWrapper()?
